pubsub/iterator.go

```diff
@@ -249,6 +252,9 @@ func (it *messageIterator) receive(maxToPull int32) ([]*Message, error) {
 	}
 	// Any error here is fatal.
 	if err != nil {
+		if status.Code(err) == codes.Canceled {
```

Comment:
On second thought, I think we shouldn't change this. Changing this is technically a breaking change, even though it's fairly unlikely. Did you have a strong reason to add this? I couldn't tell from an initial pass how this is related to the memory leak.

Reply:
No it’s indeed just that it wasn’t clear to me whether the caller or the function had the responsibility of handling ‘cancelled’ error: I was under the impression, given that cancelling the provided context is the intended way to stop receiving messages, that Receive should then just return nil, but feel free to adjust/remove this as you see fit. 

From what you were saying about unit tests failing, sounds like this change here is not enough for that anyway…

Reply:
I think if I were to go back and redesign this, `Receive` should return nil, but given that we have returned `Canceled` up until now, we should leave it as is.

Reply:
Right, sorry, looked at it more closely now: I _did_ introduce this change because without it the behaviour _does_ also change, as previously `recvMessages` would not return a cancellation error when terminating the `Receive` call, while with the suggested change it would (and the error would propagate back).

I guess the point is my fix in the error check here is a bit too blunt as it introduces two _other_ changes (presumably the source of failing tests):
1. Seems like `pullMessages` in the `it.po.synchronous==true` branch _is_ expected to return cancellation errors in some situations?
2. The `recvMessages` call currently _does_ return a cancellation error if it's the undelying client that gets closed (rather than just the context passed to `Receive`), which gets tested in `TestStreamingPull_ClosedClient`.

If you want to preserve this exact same behaviour, the fix would thus be
```
--- a/pubsub/iterator.go
+++ b/pubsub/iterator.go
@@ -249,12 +249,12 @@ func (it *messageIterator) receive(maxToPull int32) ([]*Message, error) {
                rmsgs, err = it.pullMessages(maxToPull)
        } else {
                rmsgs, err = it.recvMessages()
+               if status.Code(err) == codes.Canceled && it.ps.ctx.Err() == context.Canceled {
+                       err = io.EOF
+               }
        }
        // Any error here is fatal.
        if err != nil {
-               if status.Code(err) == codes.Canceled {
-                       err = io.EOF
-               }
                return nil, it.fail(err)
        }
        recordStat(it.ctx, PullCount, int64(len(rmsgs)))
```

Reply:
Technically [synchronous mode is deprecated](https://github.com/googleapis/google-cloud-go/blob/aaf7fa23b21cafb892ea1195b1a6e796c3d139de/pubsub/subscription.go#L926), but yes let's avoid making breaking changes there regardless.

Reply:
OK, well, I've pushed the change above, along with a comment. Let me know if you're happy with it!